### ISSUE TITLE
[css-properties-values-api] Are fallbacks provided for registered properties validated by the CP syntax?

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#10455]
Take a look at this codepen, which is surprisingly interoperable across UAs: https://codepen.io/leaverou/pen/oNRpEGO

`--color` is defined as:

```css
@property --color {
	syntax: "lime | yellowgreen";
	initial-value: yellowgreen;
	inherits: false;
}
```

| Actual | Expected |
|--------|--------|
| <img width="574" alt="image" src="https://github.com/w3c/csswg-drafts/assets/175836/3efe4fa0-00b4-4956-9385-e4aa42acc821"> | <img width="497" alt="image" src="https://github.com/w3c/csswg-drafts/assets/175836/68d4d378-28f8-437f-9f7b-8bc8cb27a489">

Given that…
1. Registered properties become IACVT when they are given a value that does not parse with their defined syntax, 
2. IACVT properties behave as if their value is `unset` and properties that do not inherit handle `unset` as `intiial` 
3. `--color`’s initial value is `yellowgreen`
4. `--color` is defined to not inherit

…I would have expected `yellowgreen`, not `transparent`. It is especially weird that this only happens when a fallback value is provided that does *not* match the property syntax.

Given the behavior is interoperable, @kizu assumed there must be a WPT test about this, but could not find a test that would test this interaction. The only place he found that tests the fallback not applying when the variable uses its fallback value is here: https://github.com/web-platform-tests/wpt/blob/master/css/css-pseudo/highlight-cascade/highlight-cascade-003.html#L8-L14
But there is no test (or we can't find it) that covers this weird behavior.

I guess, if anything describes what should happen, then it is https://drafts.csswg.org/css-variables/#invalid-variables

> Either the property’s inherited value or its initial value depending on whether the property is inherited or not, respectively, as if the property’s value had been specified as the unset keyword.

However, that is not what is happening in the second test. Changing whether the property inherits and setting `--color: lime` on the parent does not change the outcome of the second test, only the others: https://codepen.io/leaverou/pen/oNRpEGO

It should be clarified in the specs whether the value of the fallback used in the `var()` should also be validated by the custom property syntax or treated as an unregistered token stream that is validated at the point of usage (I would vote for the latter, but no strong opinion) and what the behavior is in that case.

Also tagging as Agenda+ since `@property` usage is going to pick up soon now that Firefox has implemented, and authors are soon going to start hitting this.

_Thanks to @kizu for his help in tracking this down_

**Comment by Loirooriol** [comment]
`gold` is provided as a fallback in a declaration for `background`, not for `--color`, so validating it against the grammar of `--color` seems very unexpected to me.

**Comment by tabatkins** [comment]
Yeah, fallback in var() should be completely unrelated to whatever conditions are imposed on the custom property *referenced* by the var(). Nothing in any of the specs suggests there should be such an effect; the behavior of the fallback is indeed just "an unregistered token stream that is validated at the point of usage".

**Comment by andruud** [comment]
> Nothing in any of the specs suggests there should be such an effect

It's specified here: https://drafts.css-houdini.org/css-properties-values-api-1/#fallbacks-in-var-references

See also https://github.com/w3c/css-houdini-drafts/issues/360.

---

But I agree that this seems odd now that we have other arbitrary substitution functions that can violate the "grammar" in the fallback, e.g.:

```
width: attr(my-width px, auto);
                         ^^^^ <-- Not px, but valid
```

**Comment by tabatkins** [comment]
Well dang, I'm dumb. But yeah, I think it's better to remove the restriction and let the fallback be anything.

**Comment by SebastianZ** [comment]
@miragecraft Please see @Loirooriol's [comment](https://github.com/w3c/csswg-drafts/issues/10455#issuecomment-2174259218)! The fallback value is checked against the property it is _applied to_, not the custom property in the `var()` construct.

> In the codepen: https://codepen.io/leaverou/pen/oNRpEGO
> 
> `<pre style="--color: red; background: var(--color, lime);">\</pre\>` is rendered as yellowgreen, even though lime is a valid syntax of the registered `--color` property.
> 
> The WPT test mentioned also confirms this: https://github.com/web-platform-tests/wpt/blob/master/css/css-pseudo/highlight-cascade/highlight-cascade-003.html#L8-L14
> 
> <style>
>   @property --bg {
>     syntax: "<color>";
>     inherits: false;
>     initial-value: green;
>   }
>   :root::selection {
>     background-color: var(--bg, red);
>   }
> </style>
> <main>PASS if background-color is green when selected</main>
> 
> Why would it be green if fallback value of red is a valid `<color>`?

In those two cases, the _initial value_ of the custom properties is taken, as it is a valid `<color>` value. So the fallback values are disregarded there.

Put the other way round, if you have this:

```css
@property --foo {
  syntax: "<length>";
  inherits: false;
  initial-value: 100px;
}

:root {
  background-color: var(--foo, yellowgreen);
}
```

the background should be yellowgreen because the `--foo` property is invalid at computed value time while the fallback value `yellowgreen` is a valid value for `background-color`.
At least that's what was resolved on. Though browsers currently fail to apply the fallback value because they validate it against `--foo`'s `<length>` syntax.

Sebastian

**Comment by Loirooriol** [comment]
@SebastianZ I don't think that's right, `--foo` isn't invalid at computed value time: if you aren't setting it, it has initial value of `100px`. So you get `background-color: 100px`, which makes `background-color` invalid at computed value time and thus you get a transparent background. You would need `background-color: first-valid(var(--foo), yellowgreen)` to get the desired effect.

> If no fallback would ever be applied even if they pass syntax check then what's the point of dropping validation?

@miragecraft Well, what would be the point of making the declaration invalid at computed time just because the fallback value (that will be ignored) would be invalid if you were to use it on the property that you are substituting?
